File: src/interfaces/api/schemas/platforms.py
"""
Platform API Schemas
Self-documenting schemas using type annotations and minimal descriptions.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

# ============================================================================
# Domain Types
# ============================================================================

PlatformId = UUID
OrganizationId = str
PlatformSlug = str
PlatformType = str  # 'opendatasoft' | 'datagouv'
SyncStatus = str  # 'success' | 'failed' | 'pending'
URL = str
APIKey = str

# ============================================================================
# Sync History
# ============================================================================


class PlatformSync(BaseModel):
    """Execution record of a single synchronization."""

    platform_id: PlatformId
    timestamp: datetime
    status: SyncStatus
    datasets_count: int


# ============================================================================
# Platform DTOs
# ============================================================================


class PlatformDTO(BaseModel):
    """Complete platform configuration and current state."""

    # Identity
    id: PlatformId
    name: str
    slug: PlatformSlug
    type: PlatformType

    # Configuration
    url: URL
    organization_id: OrganizationId
    key: APIKey | None = None  # Masked in public responses

    # State & Metrics
    datasets_count: int = 0
    last_sync: datetime | None = None
    last_sync_status: SyncStatus | None = None
    created_at: datetime | None = None

    # History
    syncs: list[PlatformSync] | None = None

    model_config = {"arbitrary_types_allowed": True}


class PlatformCreateDTO(BaseModel):
    """Input data for registering a new platform."""

    name: str
    slug: PlatformSlug
    organization_id: OrganizationId
    type: PlatformType
    url: URL
    key: APIKey


class PlatformCreateResponse(BaseModel):
    """Response returned after successful platform registration."""

    id: PlatformId
    name: str
    slug: PlatformSlug
    type: PlatformType
    url: URL
    key: APIKey

    model_config = {"arbitrary_types_allowed": True}


# ============================================================================
# API Responses
# ============================================================================


class PlatformsResponse(BaseModel):
    """Paginated or grouped list of platforms."""

    platforms: list[PlatformDTO]
    total_platforms: int

    model_config = {"arbitrary_types_allowed": True}

File: front/src/types/platforms.ts
// ============================================================================
// Platform Types
// ============================================================================

// ----------------------------------------------------------------------------
// Domain Primitives
// ----------------------------------------------------------------------------

type UUID = string;
type ISO8601Date = string;
type URL = string;
type PlatformSlug = string;
type APIKey = string;

// ----------------------------------------------------------------------------
// Platform Types
// ----------------------------------------------------------------------------

/**
 * Platform type determines which adapter to use for data extraction.
 * Each type has its own API contract and data structure.
 */
type PlatformType = 'opendatasoft' | 'datagouv' | string;

type SyncStatus = 'pending' | 'success' | 'failed';

// ----------------------------------------------------------------------------
// Sync History
// ----------------------------------------------------------------------------

/**
 * Record of a single synchronization execution.
 * Tracks when sync ran, its outcome, and how many datasets were processed.
 */
export type PlatformSync = {
    id: UUID;
    platform_id: UUID;
    timestamp: ISO8601Date;
    status: SyncStatus;
    datasets_count: number;
}

// ----------------------------------------------------------------------------
// Platform Reference
// ----------------------------------------------------------------------------

/**
 * Complete platform configuration and state.
 * Used for platform management and monitoring.
 */
export type PlatformRef = {
    id: UUID;
    name: string;
    slug: PlatformSlug;
    type: PlatformType;
    url: URL;
    key: APIKey;  // Masked on client side
    created: ISO8601Date;
    lastSync: ISO8601Date;
    lastSyncStatus: SyncStatus;
    datasetsCount: number;
    syncs?: PlatformSync[];  // Full sync history (optional)
};
